daktari/check_utils.py

```diff
@@ -0,0 +1,9 @@
+from typing import Union, Type, List
+
+from daktari.check import Check
+
+
+def get_all_dependent_check_names(check: Union[Check, Type[Check]]) -> List[str]:
+    dependents = [get_all_dependent_check_names(dep) for dep in check.depends_on]
+    flat_dependents = [item for sublist in dependents for item in sublist]
+    return flat_dependents + [check.name]
```

Comment:
Infinite loop is definitely possible atm, but if it were to be written I think it would cause us problems elsewhere. I'll test what happens - assuming that python blows up with a reasonable enough message (StackOverflow type thing) then that's probably good enough. 

*Maybe* we could write a test which populated one of every check via reflection and then checked there were no dependency cycles, but that sounds hard / maybe overkill :sweat_smile:

Reply:
And yeah, combining with the `check_sorter` code was my initial plan but it looked kinda non-trivial and then I forgot about it. I'll take another look :+1:

Reply:
> Maybe we could write a test which populated one of every check via reflection and then checked there were no dependency cycles, but that sounds hard / maybe overkill 

We might also need something at run time, since users can create their own checks and we can't test against that. But yeah it's not essential, just a nice to have.

Reply:
I've made it throw a nicer error - I think that's good enough as a "runtime check": 

```
$ PYTHONPATH=~/daktari python3 -m daktari
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/alyssa/daktari/daktari/__main__.py", line 55, in <module>
    sys.exit(main())
  File "/home/alyssa/daktari/daktari/__main__.py", line 21, in main
    config = read_config(args.config_path)
  File "/home/alyssa/daktari/daktari/config.py", line 38, in read_config
    return apply_local_config(config)
  File "/home/alyssa/daktari/daktari/config.py", line 55, in apply_local_config
    return remove_ignored_checks(config, ignored_checks)
  File "/home/alyssa/daktari/daktari/config.py", line 65, in remove_ignored_checks
    ignored_checks = list(filter(lambda check: check_should_be_ignored(check, ignored_check_names), config.checks))
  File "/home/alyssa/daktari/daktari/config.py", line 65, in <lambda>
    ignored_checks = list(filter(lambda check: check_should_be_ignored(check, ignored_check_names), config.checks))
  File "/home/alyssa/daktari/daktari/config.py", line 71, in check_should_be_ignored
    dependents = get_all_dependent_check_names(check)
  File "/home/alyssa/daktari/daktari/check_utils.py", line 12, in get_all_dependent_check_names
    return _get_all_dependent_check_names_recursive(check, set())
  File "/home/alyssa/daktari/daktari/check_utils.py", line 22, in _get_all_dependent_check_names_recursive
    sub_dependents = [_get_all_dependent_check_names_recursive(dep, prev_parents) for dep in check.depends_on]
  File "/home/alyssa/daktari/daktari/check_utils.py", line 22, in <listcomp>
    sub_dependents = [_get_all_dependent_check_names_recursive(dep, prev_parents) for dep in check.depends_on]
  File "/home/alyssa/daktari/daktari/check_utils.py", line 20, in _get_all_dependent_check_names_recursive
    raise CyclicCheckException(f"Check [{intersection.pop()}] has cyclic dependencies")
daktari.check_utils.CyclicCheckException: Check [intellij.installed] has cyclic dependencies
```

Turns out this is pretty hard to arrange anyway, even if you're trying - you have to do it via mutation because Python doesn't let you reference a class before it's declared. Oh well, it's handled now :grin: